src/semver/version.py

```diff
@@ -92,21 +110,81 @@ class Version:
+            Optional[StringOrInt],  # build
+        ],
+        major: SupportsInt = 0,
         minor: SupportsInt = 0,
         patch: SupportsInt = 0,
-        prerelease: Union[String, int] = None,
-        build: Union[String, int] = None,
+        prerelease: StringOrInt = None,
+        build: StringOrInt = None,
     ):
+        def _check_types(*args):
+            if args and len(args) > 5:
+                raise ValueError("You cannot pass more than 5 arguments to Version")
+            elif len(args) > 1 and "." in str(args[0]):
+                raise ValueError(
+                    "You cannot pass a string and additional positional arguments"
+                )
+            allowed_types_in_args = (
+                (int, str, bytes),  # major
+                (int, str, bytes),  # minor
+                (int, str, bytes),  # patch
+                (str, bytes, int, type(None)),  # prerelease
+                (str, bytes, int, type(None)),  # build
+            )
+            return [
+                isinstance(item, allowed_types_in_args[i])
+                for i, item in enumerate(args)
+            ]
+
+        cls = self.__class__
+        verlist: List[Optional[StringOrInt]] = [None, None, None, None, None]
+
+        types_in_args = _check_types(*args)
+        if not all(types_in_args):
+            pos = types_in_args.index(False)
+            raise TypeError(
+                "not expecting type in argument position "
+                f"{pos} (type: {type(args[pos])})"
+            )
+        elif args and "." in str(args[0]):
+            # we have a version string as first argument
+            v = cls._parse(args[0])  # type: ignore
+            for idx, key in enumerate(
+                ("major", "minor", "patch", "prerelease", "build")
+            ):
+                verlist[idx] = v[key]
+        else:
+            for index, item in enumerate(args):
+                verlist[index] = args[index]  # type: ignore
+
         # Build a dictionary of the arguments except prerelease and build
-        version_parts = {"major": int(major), "minor": int(minor), "patch": int(patch)}
+        try:
+            version_parts = {
+                # Prefer major, minor, and patch arguments over args
+                "major": int(major or verlist[0] or 0),
+                "minor": int(minor or verlist[1] or 0),
+                "patch": int(patch or verlist[2] or 0),
+            }
+        except ValueError:
+            raise ValueError(
+                "Expected integer or integer string for major, minor, or patch"
+            )
 
         for name, value in version_parts.items():
             if value < 0:
                 raise ValueError(
                     "{!r} is negative. A version can only be positive.".format(name)
                 )
 
+        prerelease = cls._ensure_str(prerelease or verlist[3])  # type: ignore
+        build = cls._ensure_str(build or verlist[4])  # type: ignore
```

Comment:
Here is the proper way to use `_ensure_str`:
```suggestion
        if isinstance(prerelease, int):
            self._prerelease = prerelease
        else:
            self._prerelease = cls._ensure_str(prerelease or verlist[3])
        if isinstance(build, int):
            self._build= build
        else:
            self._build = cls._ensure_str(build or verlist[4])
```

Reply:
You mean this should be ignored, right?

Reply:
Nope that's the good one. The first review ( the one where I suggested to change the type hints of `_ensure_str`) is the one that's bad. So we should change it back to what it was and add this in to make the types correct by checking if it is an int.

Reply:
I'll put a sad face on the bad reviews

Reply:
Ahh, ok, thanks for the clarification. :+1: 

I'm a bit confused. Shouldn't prerelease and build only be None or str? At least our property say so:

```python
@property
def prerelease(self) -> Optional[str]:
```

So this line would probably raise errors in mypy:

```python
if isinstance(prerelease, int):
            self._prerelease = prerelease
```

Is my assumption correct?

Reply:
Well I want to consider also the fact that somebody might not want to provide the format but only the relevant information (the number). Although maybe we should keep that for later and simply consider the case of a string for now.